// allere-constants/src/question-type.js
/**
 * Created by 彭建伦 on 2016/9/1.
 * 试题类型
 */
export const SOLVE_PROBLEM_LEGACY = 'application';
export const FILL_BLANKS = "fillblanks";//填空题     // 填空(fillblanks)   ，候选得分点类型<standalone, listSetItems, enumSetItems>
export const VERTICAL_EXP_CALC = "vertical-exp-calc";     // 竖式计算(vertical-exp-calc)   ，候选得分点类型<vertical>
export const VERTICAL_EXP_ERROR = "vertical-exp-error";     // 竖式改错(vertical-exp-error)  ，候选得分点类型<vertical>
export const VERTICAL_EXP_BLANK = "vertical-exp-blank";     // 竖式填空(vertical-exp-blank)  ，候选得分点类型<vertical>
export const OPTIONS = "options";     // 选择(options)   ，候选得分点类型<standalone(仅适用单选), selection(单选或多项选择)>
export const JUDGEMENT = "judgement";     // 判断(judgement)  ，候选得分点类型<standalone>
export const SOLVE_PROBLEM = "application";     // 解决问题(solve-problem)   ，候选得分点类型<application>
export const SOLVE_VARIABLE = "variable-solve";     // 解方程(solve-variable)  ，候选得分点类型<variable-solve>
export const PULL_EXP_CALC = "pull-exp";     // 脱式计算(pull-exp-calc)  ，候选得分点类型<pull-exp>
export const PULL_EXP_CALC_LEGACY = "pull-exp"; //老题库的脱式计算类型
